### PR TITLE
Propagate more compile benchmark metadata into the dashboard

### PR DESCRIPTION
**Pull request by Kobzol**
This PR adds a **new required field to perf-config.json**, `artifact`. It has to be either `library` or `binary`, it specifies which target does the benchmark build. There's now a unit test that checks that all `perf-config.json` contain this attribute and that they be correctly parsed. Before this would only be found out during actual benchmarking or when displaying the compare page.

Furthermore, information about compile-time benchmarks that is useful (e.g. did the benchmark override LTO/CGU/debug information? does it build a binary or a library?) is now propagated into the compare page dashboard.

The information is extracted from the compile benchmark directory (and its `perf-config.json` and `Cargo.toml` files) via a build script of the `site` package. It is then sent to the compage page via the API and displayed as a simple tooltip over each benchmark name in the comparison table. The build script takes ~0.05s to execute, and is cached unless something in `collector/compile-benchmarks` changes (it seems that the detection whether something has changed is quite fast and doesn't slow down `site` rebuilds).

In addition, the filters can now hide/show all binaries/all libraries. This filtering is implemented using the newly added metadata.

perf

Fixes: https://github.com/rust-lang/rustc-perf/issues/1619

### CODE REVIEW

**Review by lqd — COMMENTED** [review]
This LGTM but I'm not familiar with:
- some parts of the collector/api
- using serde in real life, and the intricacies of missing data, backwards compatibility in format, deserialization errors, etc

so maybe another quick look from someone more familiar with these could be worthwhile

File: site/src/benchmark_metadata/mod.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    metadata
+        .benchmarks
+        .into_iter()
+        .map(|(name, metadata)| {
+            let metadata::CompileBenchmarkMetadata {
+                perf_config,
+                release_metadata,
+            } = metadata;
+            let perf_config: BenchmarkConfig =
+                serde_json::from_value(perf_config).expect("Cannot deserialize perf-config.json");
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply site/src/benchmark_metadata/mod.rs:55]: I'm a bit worried that if we leave it optional, it will just silently be set to a wrong value. But you raise a good point. The error message should be improved, and I should also modify `collector download` to automatically generate the artifact type.

[the remark this answers is not in the document]
Comment (lqd) [reply site/src/benchmark_metadata/mod.rs:55]: If we only use it in the tooltip, it's probably fine. The filtering may be more impactful, and we could show missing data present in both lib/bin categories (with the help text explaining how to fix it). Though I'm not sure if using the filter would be very common, as we have basically 2 binaries and a bunch of simili-hello world.

**Comment by Kobzol** [comment]
TLDR: This PR adds a new required field to `perf-config.json` of each compilation benchmark (whether it builds a library or a binary). The perf configs are now loaded in a build script of `site`. To avoid depending on the `BenchmarkConfig` struct, which is the in-memory representation of that JSON, the build script loads the file as opaque JSON, and it is only deserialize in the website on-demand, when the compare page is loaded. This means that if someone creates a new benchmark and forgets to set the field, CI probably won't catch this currently. The `collector download` command, which should be used to add new benchmarks, will require setting this though (once I implement it). Furthermore, if the new benchmark won't be benchmarked on CI, it will also crash on the perf. server. An alternative could be to include the file containing `BenchmarkConfig` from `collector` into the build script, but it seems a bit fragile to include the file with `#[path] mod` like this across crates.

However, we already have a precedent for this - we added a required field `category` before.

The field is required to silently avoid having wrong values there. That being said, it has some problems mentioned above, and having the field "mislabeled" doesn't really cause any high risk. So making it optional (and default to `library`, which is much more common), could be a safer solution.

I think that a best solution would be to keep it required, and introduce some new CI check that will make sure that all benchmarks have proper JSON configurations.

**Comment by Kobzol** [comment]
I added a simple unit test that checks that all `perf-config.json` files can be deserialized. This shold resolve the issue with invalid configuration files.

**Comment by Kobzol** [comment]
The frontend side was reviewed by @lqd, also requesting review from @Mark-Simulacrum about the config changes and the propagation of data into `site`.

**Comment by nnethercote** [comment]
Do [these instructions](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks#add-a-new-benchmark) need updating?

**Comment by Kobzol** [comment]
Good point! I updated it, along with the glossary.

**Review by Mark-Simulacrum — APPROVED** [review]
